### PR TITLE
[PRW 2.0] Added Sender and RW Handler support for Response Stats.

### PR DESCRIPTION
**Pull request by bwplotka**
Chained on top of https://github.com/prometheus/prometheus/pull/14427

Fixes https://github.com/prometheus/prometheus/issues/14359

### CODE REVIEW

File: storage/remote/client.go [hunk 1]
```diff
@@ -267,28 +267,38 @@ func (c *Client) Store(ctx context.Context, req []byte, attempt int) error {
...
-		err = fmt.Errorf("server returned HTTP status %s: %s", httpResp.Status, line)
+	if httpResp.StatusCode/100 == 2 {
+		return rs, nil
+	}
+
+	// Handling errors e.g. read potential error in the body.
+	scanner := bufio.NewScanner(io.LimitReader(httpResp.Body, maxErrMsgLen))
+	line := ""
+	if scanner.Scan() {
+		line = scanner.Text()
```
Comment (TheSpiritXIII) [inline storage/remote/client.go:292]: question: Could this not get the full error message? Should we loop and add append more text, or maybe put an indication that we didn't get the whole thing (for example with ellipsis)?

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -1661,33 +1662,43 @@ func populateTimeSeries(batch []timeSeries, pendingData []prompb.TimeSeries, sen
...
 	if err != nil {
-		level.Error(s.qm.logger).Log("msg", "non-recoverable error", "count", sampleCount, "exemplarCount", exemplarCount, "histogramCount", histogramCount, "err", err)
-		s.qm.metrics.failedSamplesTotal.Add(float64(sampleCount))
-		s.qm.metrics.failedExemplarsTotal.Add(float64(exemplarCount))
-		s.qm.metrics.failedHistogramsTotal.Add(float64(histogramCount))
+		level.Error(s.qm.logger).Log("msg", "non-recoverable error", "failedSampleCount", sampleCount-rs.Samples, "failedHistogramCount", histogramCount-rs.Histograms, "failedExemplarCount", exemplarCount-rs.Exemplars, "err", err)
+	}
+
+	// Partial errors may happen -- account for that.
+	// TODO(bwplotka): No error and non zero diffs might mean something weird happens e.g. buggy 2.0 Receiver. Warn?
```
Comment (TheSpiritXIII) [inline storage/remote/queue_manager.go:1685]: suggestion: I think that's a great idea, and you already have a logger. What's stopping you?

File: storage/remote/write_handler.go [hunk 3]
```diff
@@ -324,16 +324,85 @@ const (
...
+}
+
+// AllSamples returns both float and histogram sample numbers.
+func (s WriteResponseStats) AllSamples() int {
+	return s.Samples + s.Histograms
+}
+
+// Add returns the sum of this WriteResponseStats plus the given WriteResponseStats.
+func (s WriteResponseStats) Add(rs WriteResponseStats) WriteResponseStats {
+	s.Confirmed = rs.Confirmed
```
Comment (TheSpiritXIII) [inline storage/remote/write_handler.go:355]: nit: I'm not sure this looks right. Should we maybe set it to false when we add at least one that is false? Or maybe if that's not possible to add a false one, be clear and always set to true?

**Review by TheSpiritXIII — COMMENTED** [review]
Partial review (because I'm not too familiar with the code) to help Bartek start some discussions.

File: storage/remote/queue_manager.go [hunk 4]
```diff
@@ -1661,33 +1662,43 @@ func populateTimeSeries(batch []timeSeries, pendingData []prompb.TimeSeries, sen
 
 func (s *shards) sendSamples(ctx context.Context, samples []prompb.TimeSeries, sampleCount, exemplarCount, histogramCount int, pBuf *proto.Buffer, buf *[]byte, enc Compression) error {
 	begin := time.Now()
-	err := s.sendSamplesWithBackoff(ctx, samples, sampleCount, exemplarCount, histogramCount, 0, pBuf, buf, enc)
-	s.updateMetrics(ctx, err, sampleCount, exemplarCount, histogramCount, 0, time.Since(begin))
+	rs, err := s.sendSamplesWithBackoff(ctx, samples, sampleCount, exemplarCount, histogramCount, 0, pBuf, buf, enc)
+	s.updateMetrics(ctx, err, sampleCount, exemplarCount, histogramCount, 0, rs, time.Since(begin))
```
Comment (cstyan) [inline storage/remote/queue_manager.go:1666]: I like that we would be capturing and using the stats if they exist for a v1 request. We should be sure we document this change clearly in the changelog, just to avoid any confusion if someone's still using v1 and then all of a sudden their failed datapoints metrics start to look very different because their receiver has started supporting the response stat headers for v1.

File: storage/remote/queue_manager.go [hunk 5]
```diff
@@ -1850,15 +1890,27 @@ func (s *shards) sendV2SamplesWithBackoff(ctx context.Context, samples []writev2
...
+		addStats(rs)
 
-		if err != nil {
-			span.RecordError(err)
-			return err
+		if err == nil {
+			// Check the case mentioned in PRW 2.0 https://github.com/prometheus/docs/pull/2486
+			if sampleCount+histogramCount+exemplarCount > 0 && rs.NoDataWritten() {
+				err = fmt.Errorf("sent v2 request with %v samples, %v histograms and %v exemplars; got 2xx, but PRW 2.0 response header statistics indicate %v samples, %v histograms and %v exemplars were accepted;"+
+					" assumining failure e.g. the target only supports PRW 1.0 prometheus.WriteRequest, but does not check the Content-Type header correctly",
```
Comment (cstyan) [inline storage/remote/queue_manager.go:1903]: should we include a link to that part of the 1.0 spec here?

File: storage/remote/write_handler.go [hunk 6]
```diff
@@ -324,16 +324,85 @@ const (
 	rw20WrittenExemplarsHeader  = "X-Prometheus-Remote-Write-Written-Exemplars"
 )
 
-type responseStats struct {
-	samples    int
-	histograms int
-	exemplars  int
+// WriteResponseStats represents the response write statistics specified in https://github.com/prometheus/docs/pull/2486
```
Comment (cstyan) [inline storage/remote/write_handler.go:327]: nit, just feels like these stats structs should exist in a different file than the write handler code :shrug:

File: storage/remote/write_handler.go [hunk 7]
```diff
@@ -453,26 +522,26 @@ func (h *writeHandler) appendV2(app storage.Appender, req *writev2.Request, rs *
...
 			if errors.Is(err, storage.ErrOutOfOrderExemplar) {
-				outOfOrderExemplarErrs++
-				level.Debug(h.logger).Log("msg", "Out of order exemplar", "err", err.Error(), "series", ls.String(), "exemplar", fmt.Sprintf("%+v", e))
+				outOfOrderExemplarErrs++ // Maintain old metrics, but technically not needed, given we fail here.
+				level.Error(h.logger).Log("msg", "Out of order exemplar", "err", err.Error(), "series", ls.String(), "exemplar", fmt.Sprintf("%+v", e))
+				badRequestErrs = append(badRequestErrs, fmt.Errorf("%w for series %v", err, ls.String()))
 				continue
 			}
-			level.Debug(h.logger).Log("msg", "Error while adding exemplar in AppendExemplar", "series", ls.String(), "exemplar", fmt.Sprintf("%+v", e), "err", err)
+			// TODO(bwplotka): Are we confident enough to 500 on exemplars in Prometheus?
```
Comment (cstyan) [inline storage/remote/write_handler.go:535]: I don't think the exemplar storage has really been touched since I first wrote it, so I would lean towards no, but I could be convinced otherwise by anyone who has worked on the exemplar storage since then.

Comment (bwplotka) [reply storage/remote/write_handler.go:535]: Do we see a lot of internal errors? We can take easier path and keep the internal errors silent for now 💪🏽

File: storage/remote/queue_manager.go (hunk 2 above)
Comment (bwplotka) [reply storage/remote/queue_manager.go:1685]: hope that this will not happen 💣  xD

File: storage/remote/write_handler.go (hunk 3 above)
Comment (bwplotka) [reply storage/remote/write_handler.go:355]: Yea, this is tricky. I am not sure how retries should work with those stats, it depends if on the receiver independence etc. Shouldn't be a big deal too, we might never got it right, but some consistency would be nice.

My thinking is that, if it was once confirmed it's likely it's generally 2.0 Receiver, so all is confirmed? Maybe `confirm` is a too generic name and we should actually name it `likely2.0` or something

File: cmd/promtool/metrics.go [hunk 8]
```diff
@@ -116,7 +116,7 @@ func parseAndPushMetrics(client *remote.Client, data []byte, labels map[string]s
 
 	// Encode the request body into snappy encoding.
 	compressed := snappy.Encode(nil, raw)
-	err = client.Store(context.Background(), compressed, 0)
+	_, err = client.Store(context.Background(), compressed, 0)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply cmd/promtool/metrics.go:119]: I don't know what's even that command is doing, but yes =D

File: storage/remote/client.go [hunk 9]
```diff
@@ -267,28 +267,38 @@ func (c *Client) Store(ctx context.Context, req []byte, attempt int) error {
...
 	}
 	defer func() {
 		io.Copy(io.Discard, httpResp.Body)
 		httpResp.Body.Close()
 	}()
 
+	rs, err := ParseWriteResponseStats(httpResp)
+	if err != nil {
+		// There were some response header values we can't parse, so fail fast here.
+		return WriteResponseStats{}, err
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/remote/client.go:280]: Correct. I decided to not trust any of the headers if once header was malformed - otherwise we might have a skewed observation 🤔 But maybe we would fail request we could be otherwise valid? Not sure.

File: storage/remote/write_handler.go (hunk 6 above)
Comment (bwplotka) [reply storage/remote/write_handler.go:327]: agree, will change

File: storage/remote/client.go (hunk 1 above)
Comment (bwplotka) [reply storage/remote/client.go:292]: good point, will do

File: storage/remote/queue_manager.go (hunk 4 above)
Comment (bwplotka) [reply storage/remote/queue_manager.go:1666]: Added to the changelog!

File: storage/remote/queue_manager.go (hunk 2 above)
Comment (bwplotka) [reply storage/remote/queue_manager.go:1685]: fixed, great point

File: storage/remote/queue_manager.go [hunk 10]
```diff
@@ -1780,29 +1803,46 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
...
 		// Failing to build the write request is non-recoverable, since it will
 		// only error if marshaling the proto to bytes fails.
-		return err
+		return WriteResponseStats{}, err
 	}
 
 	reqSize := len(req)
 	*buf = req
 
+	totalRs := WriteResponseStats{}
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/remote/queue_manager.go:1838]: done, thanks

File: storage/remote/queue_manager.go (hunk 5 above)
Comment (bwplotka) [reply storage/remote/queue_manager.go:1903]: Only if it's merged though (:

File: storage/remote/write_handler.go (hunk 6 above)
Comment (cstyan) [reply storage/remote/write_handler.go:327]: The write client is using them as well?

Separate file is the only one that makes sense, otherwise they're either in the cient or write handler :man_shrugging: not a big deal to stick with this or change imo.

File: storage/remote/write_handler.go (hunk 7 above)
Comment (cstyan) [reply storage/remote/write_handler.go:535]: I actually don't know tbh, I don't use them much myself, so I'd defer to anyone who more experience with them. I can ask our internal platform team or Mimir next week.

File: storage/remote/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cstyan) [reply storage/remote/client.go:279]: I feel like using the results if any were present is valid. If it isn't, we might as well modify the `ParseWriteResponseStats` to fail fast and return the first error instead of attempting to parse all headers?

File: cmd/promtool/metrics.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cstyan) [reply cmd/promtool/metrics.go:119]: I think it's meant for testing remote write receivers, can take a text file and push it's content as prometheus remote write: https://prometheus.io/docs/prometheus/latest/command-line/promtool/#promtool-push

File: storage/remote/write_handler.go (hunk 3 above)
Comment (cstyan) [reply storage/remote/write_handler.go:355]: @TheSpiritXIII not sure I understand your comment, add one of _what_ that is false?

File: storage/remote/write_handler.go (hunk 7 above)
Comment (cstyan) [reply storage/remote/write_handler.go:535]: Got some context here from @bboreham. Mimir recently moved from rejecting data that contained invalid exemplars to only rejecting the invalid exemplars themselves [here](https://github.com/grafana/mimir/pull/8224). I think we should default to doing the same, and if there's an ask for it in the future we could add a `strict` option and consider changing the default.

File: CHANGELOG.md [hunk 11]
```diff
@@ -2,13 +2,16 @@
...
+* [ENHANCEMENT] Remote-Write: 1.x messages against Remote Write 2.x Receivers will have now correct values for `prometheus_storage_<samples|histograms|exemplar>_failed_total` in case of partial errors #14444
+
 ## 2.53.1 / 2024-07-10
 
 Fix a bug which would drop samples in remote-write if the sending flow stalled
 for longer than it takes to write one "WAL segment". How long this takes depends on the size
 of your Prometheus; as a rough guide with 10 million series it is about 2-3 minutes.
 
-* [BUGFIX] Remote-write: stop dropping samples in catch-up #14446
+* [BUGFIX] Remote-Write: stop dropping samples in catch-up #14446
```
Comment (bboreham) [inline CHANGELOG.md:14]: Unrelated change.

File: storage/remote/write_handler.go (hunk 7 above)
Comment (bwplotka) [reply storage/remote/write_handler.go:447]: Hm, not quite, sounds like Mimir still fails on some exemplar 400 failures: https://github.com/grafana/mimir/pull/8224/files#r1684633427

Nevertheless 400 fail never stops sample ingestion, so we are good here? I'm more concerned about 500 🤔

File: CHANGELOG.md (hunk 11 above)
Comment (bwplotka) [reply CHANGELOG.md:14]: Related, it was spelled wrongly?

File: storage/remote/queue_manager.go (hunk 5 above)
Comment (bwplotka) [reply storage/remote/queue_manager.go:1910]: Merged now, we can add.

File: CHANGELOG.md (hunk 11 above)
Comment (bwplotka) [reply CHANGELOG.md:14]: happy to kill this, you are right, it's not our item (:

**Comment by bwplotka** [comment]
Fixes https://github.com/prometheus/prometheus/issues/14323

File: cmd/promtool/metrics.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply cmd/promtool/metrics.go:119]: added TODO

File: storage/remote/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply storage/remote/client.go:279]: fair, fixing

File: storage/remote/write_handler.go (hunk 7 above)
Comment (bwplotka) [reply storage/remote/write_handler.go:447]: Addressed as discussed on Slack - keep 400 but make 500 silent for callers for now with todo for strict mode if requested by users

File: CHANGELOG.md [hunk 12]
```diff
@@ -11,7 +11,7 @@ Fix a bug which would drop samples in remote-write if the sending flow stalled
 for longer than it takes to write one "WAL segment". How long this takes depends on the size
 of your Prometheus; as a rough guide with 10 million series it is about 2-3 minutes.
 
-* [BUGFIX] Remote-Write: stop dropping samples in catch-up #14446
+* [BUGFIX] Remote-write: stop dropping samples in catch-up #14446
```
Comment (cstyan) [inline CHANGELOG.md:14]: we can still remove this

**Review by cstyan — APPROVED** [review]
we can just remove the changelog change imo

File: CHANGELOG.md (hunk 12 above)
Comment (bwplotka) [reply CHANGELOG.md:14]: hm, I thought I did, sorry

Comment (bwplotka) [reply CHANGELOG.md:14]: so, I don't think I touch this anymore in this diff, can you double check?
